### PR TITLE
[OBSDEF-5850] Feature federation changes to support listing of services and objectstores via fedsvc

### PR DESCRIPTION
**Pull request by SeemaTahaliyani**
## Purpose
[OBSDEF-5850](https://jira.cec.lab.emc.com:8443/browse/OBSDEF-5850)

- Service account changes for federation
- Service annotation changes required for listing objectscale services via fedsvc api.
- Enable federation by default
- Changed the port name to http and https to be used in creating internal and external endpoints by fedsvc.
- Supported APIs and output that works with changes in this PR.
  [SupportedFedsvcGETAPIs.txt](https://github.com/EMCECS/charts/files/5974135/SupportedFedsvcGETAPIs.txt)

## PR checklist
- [ ] make test passed
- [x] make build passed
- [x] helm install <chart> passed
- [ ] vSphere7 make package deployments passed
- [ ] helm upgrade <chart> passed
- [ ] helm test <release_name> passed

## Testing
Tested locally on minikube.

Make test is failing with same error as in master
`helm unittest ecs-cluster objectscale-manager mongoose zookeeper-operator atlas-operator decks kahm dks-testapp fio-test sonobuoy dellemc-license service-pod objectscale-graphql helm-controller objectscale-vsphere objectscale-portal objectscale-iam pravega-operator bookkeeper-operator supportassist decks-support-store statefuldaemonset-operator influxdb-operator federation logging-injector dcm
Error: fork/exec /root/.local/share/helm/plugins/helm-unittest/untt: no such file or directory
Makefile:57: recipe for target 'test' failed
make: *** [test] Error 1
`

### CODE REVIEW

File: ecs-cluster/dev.yaml [hunk 1]
```diff
@@ -394,7 +394,7 @@ managementGateway:
 
     # Specify the service port configuration for the managementGateway service
     ports:
-      - name: api
+      - name: https
```
Comment (vgupta-mickey) [inline ecs-cluster/dev.yaml:397]: is it necessary?
I am suspecting that  we have some of the code where we  search port by service name. If we change from  api to https, we need a regression on this.
```
func GetSvcNameAndPort(namespace string, releaseName string, clientset kubernetes.Interface) (string, error) {
        serviceName := releaseName + mgmtAPISvcNameSuffix
        service, err := clientset.CoreV1().Services(namespace).Get(serviceName, metav1.GetOptions{})

        if err != nil {
                log.Errorf("rest service: %s in namespace: %s is not available", serviceName, namespace)

                return "", err
        }

        port := 0
        // service listening port
        for _, entry := range service.Spec.Ports {
                if entry.Name == "api" {
                        port = int(entry.Port)
                }
        }

        addr := serviceName + "." + namespace + mgmtAPIDNSSvcNameSuffix
        hostPort := net.JoinHostPort(addr, strconv.Itoa(port))

        return hostPort, nil
}
```

File: ecs-cluster/values-harbor.yaml [hunk 2]
```diff
@@ -393,7 +393,7 @@ managementGateway:
 
     # Specify the service port configuration for the managementGateway service
     ports:
-      - name: api
+      - name: https
```
Comment (vgupta-mickey) [inline ecs-cluster/values-harbor.yaml:396]: same

File: ecs-cluster/values.yaml [hunk 3]
```diff
@@ -585,7 +585,7 @@ managementGateway:
 
     # Specify the service port configuration for the managementGateway service
     ports:
-      - name: api
+      - name: https
```
Comment (vgupta-mickey) [inline ecs-cluster/values.yaml:588]: is it necessary?
I am suspecting that we have some of the code where we search port by service name. If we change from api to https, we need a regression on this.

func GetSvcNameAndPort(namespace string, releaseName string, clientset kubernetes.Interface) (string, error) {
        serviceName := releaseName + mgmtAPISvcNameSuffix
        service, err := clientset.CoreV1().Services(namespace).Get(serviceName, metav1.GetOptions{})

        if err != nil {
                log.Errorf("rest service: %s in namespace: %s is not available", serviceName, namespace)

                return "", err
        }

        port := 0
        // service listening port
        for _, entry := range service.Spec.Ports {
                if entry.Name == "api" {
                        port = int(entry.Port)
                }
        }

        addr := serviceName + "." + namespace + mgmtAPIDNSSvcNameSuffix
        hostPort := net.JoinHostPort(addr, strconv.Itoa(port))

        return hostPort, nil
}

File: ecs-cluster/values.yaml [hunk 4]
```diff
@@ -470,10 +470,10 @@ s3:
 
     # Specify the service port configuration for the s3 service
     ports:
-      - name: secure
+      - name: https
```
Comment (vgupta-mickey) [inline ecs-cluster/values.yaml:473]: not sure about it if it needs any regression.

File: federation/README.md [hunk 5]
```diff
@@ -10,11 +10,8 @@ This Helm chart deploys a Dell EMC Federation Service and its dependencies.
 
 The Dell EMC Federation Service for the secure interaction between Objectscale instance.
 
-By default the federation service feature is not enabled in objectscale-manager.
+By default the federation service feature is enabled in objectscale-manager.
+
```
Comment (vgupta-mickey) [inline federation/README.md:14]: you can add how to disable it?

File: federation/templates/federation-admin-role.yaml [hunk 6]
```diff
@@ -0,0 +1,54 @@
+---
+{{- if or (eq .Values.global.platform "VMware") (eq .Values.global.watchAllNamespaces true) }}
+kind: ClusterRole
+{{- else }}
+kind: Role
+{{- end }}
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: {{ .Release.Namespace }}-federation-admin
+  {{- if and (ne .Values.global.platform "VMware") (eq .Values.global.watchAllNamespaces false) }}
```
Comment (vgupta-mickey) [inline federation/templates/federation-admin-role.yaml:10]: if allWatchNamespace is false, we should set the namespace, otherwise not.

File: federation/values.yaml [hunk 7]
```diff
@@ -2,6 +2,11 @@
 tag: 0.67.0
 global:
   registry: emccorp
+  # Default: true == Watch ALL namespaces
+  watchAllNamespaces: true
+  # Platform:
+  platform: Default
```
Comment (vgupta-mickey) [inline federation/values.yaml:8]: why did you add this?

**Review by vgupta-mickey — COMMENTED** [review]
couple of comments otherwise looks good.

File: ecs-cluster/dev.yaml (hunk 1 above)
Comment (lipak2345) [reply ecs-cluster/dev.yaml:397]: @vgupta-mickey, how do you decide whether to make a secure or insecure connection to the service? Maybe in the "api" case it has to be secure. It is possible that some other services expose both secure and insecure ports. 

The federation service needs a convention to figure out what type of URL to create for internal connections for services. We were thinking of denoting ports by http and https. We can even add http and https as a prefix or suffix to identify it as such. Is there any way to figure this out other than by some sort of convention.

File: federation/values.yaml (hunk 7 above)
Comment (lipak2345) [reply federation/values.yaml:8]: Will defer to @SeemaTahaliyani but I think this is needed so Federation service can get/watch object store related CRD instances and update the corresponding object store and its services information.

Comment (vgupta-mickey) [reply federation/values.yaml:8]: WatchAllNamespace and Platforms are global flags,  they should be inherited from the parent chart.  The default settings are fine but they would be overwritten by the parent chart.

File: ecs-cluster/dev.yaml (hunk 1 above)
Comment (vgupta-mickey) [reply ecs-cluster/dev.yaml:397]: I think I like http and https names but I am worried about if the internal containers are accessing these ports by name.  I have cut and pasted one piece of the code where we find out the management-gateway port by port-name.

Comment (vgupta-mickey) [reply ecs-cluster/dev.yaml:397]: as of now, there is no standard port name "http" for 80 and "https" for 443 in ECS port naming conventions. So the only way you need to figure out which port is secure or which one is not secure is by service itself only.
If managementGateway
 "api" port is https port.

For "s3"
"secure" port name is  https and 
 "insecure" is http.

Comment (lipak2345) [reply ecs-cluster/dev.yaml:397]: Why not make the change now with a common convention? We are not averse having something like "http-api" or "https-api". Either way we have to make corresponding changes for federation to be able to generate the internal URL to reach a service. 

BTW I feel that in general  other services should query federation svc for list of services and use the result to identify the internal URL location where they can be reached.

File: federation/templates/federation-admin-role.yaml (hunk 6 above)
Comment (lipak2345) [reply federation/templates/federation-admin-role.yaml:10]: It looks like namespace is set only when watchAllNamespaces is false (`eq .Values.global.watchAllNamespaces false`)

File: federation/values.yaml (hunk 7 above)
Comment (lipak2345) [reply federation/values.yaml:8]: Federation service needs to access ecsclusters so it can update object store status and information regarding its services. We have defined a new Service Account and created a ClusterRoleBinding with a  ClusterRole for this. How can we achieve that?

File: ecs-cluster/dev.yaml (hunk 1 above)
Comment (SeemaTahaliyani) [reply ecs-cluster/dev.yaml:397]: @vgupta-mickey where repo is the code that you mentioned above looks for "api" it needs to change, ideally we shouldn't be hardcoding any such names.

File: ecs-cluster/values.yaml (hunk 3 above)
Comment (SeemaTahaliyani) [reply ecs-cluster/values.yaml:588]: Yes, for fedsvc to be able to build correct  endpoints, can't add customizations for different names :)

File: ecs-cluster/values.yaml (hunk 4 above)
Comment (SeemaTahaliyani) [reply ecs-cluster/values.yaml:473]: If name is used somewhere it needs to change.

File: federation/templates/federation-admin-role.yaml (hunk 6 above)
Comment (SeemaTahaliyani) [reply federation/templates/federation-admin-role.yaml:10]: remove "and (ne .Values.global.platform "VMware")"

File: federation/README.md (hunk 5 above)
Comment (SeemaTahaliyani) [reply federation/README.md:14]: IMO fedsvc shouldn't be disable, so purposefully didn't add that.

File: federation/templates/federation-admin-role.yaml [hunk 8]
```diff
@@ -0,0 +1,54 @@
+---
+{{- if or (eq .Values.global.platform "VMware") (eq .Values.global.watchAllNamespaces true) }}
```
Comment (vgupta-mickey) [inline federation/templates/federation-admin-role.yaml:2]: you don't need this Vmware condition here as well.

**Review by vgupta-mickey — APPROVED** [review]
just one comment

File: federation/templates/federation-admin-role.yaml (hunk 8 above)
Comment (bst525) [reply federation/templates/federation-admin-role.yaml]: removed.

**Review by spiegela — APPROVED** [review]
